### DISCUSSION TITLE
"no trusted RSA public key found for" with authentication based on CA and sub-CA

### DISCUSSION THREAD

**Discussion by frmro** [strongswan/strongswan#2311] [Q&A] [answered]
Hi,

I have strongswan 5.9.5 installed on Linux Ubuntu 22.04.

I am trying to establish an ipsec tunnel with an authentication by certificate on both sides. I have the same CA and sub-CA certificates installed on both sides. Note that CA and sub-CA certificates on client and server don’t have the same dates, but subject, issuer, and public key are the same.

I have tried different configurations and always get this error: No trusted RSA public key found for

I have control over the client’s configuration, but not much on the server side.

----------
Here is the list of certificates on the client’s side:
root@ubuntupoc:~# **ipsec listcerts**
List of X.509 End Entity Certificates
  subject:  "L=WTFRT101, O=CPS, CN=C1R6G"
  issuer:   "C=CA, ST=QC, O=AAA-BBB, OU=ICP, CN=WEXI800-CA"
  validity:  not before May 14 08:49:39 2024, ok
             not after  May 14 08:59:39 2027, ok (expires in 1057 days)
  serial:    4a:00:00:41:6e:e1:76:66:75:e5:27:36:39:00:01:00:00:41:6e
  flags:     serverAuth
  authkeyId: d0:50:9b:f1:5a:09:d3:12:fc:5d:a2:75:f1:75:83:da:4e:ad:fd:bd
  subjkeyId: 86:a7:a6:b1:b9:b8:66:af:b6:ca:53:b3:81:ec:fb:df:7d:e2:50:50
  pubkey:    RSA 1024 bits, has private key
  keyid:     f5:fa:2d:93:9c:02:c1:a4:eb:a0:40:9b:b3:fc:11:7b:9f:c8:c8:a8
  subjkey:   86:a7:a6:b1:b9:b8:66:af:b6:ca:53:b3:81:ec:fb:df:7d:e2:50:50

root@ubuntupoc:~# **ipsec listcacerts**
List of X.509 CA Certificates
  subject:  "C=CA, ST=QC, O=AAA-BBB, OU=ICP, CN=WEICPACR-CA"
  issuer:   "C=CA, ST=QC, O=AAA-BBB, OU=ICP, CN=WEICPACR-CA"
  validity:  not before Mar 28 09:09:31 2019, ok
             not after  Sep 07 19:11:20 2030, ok (expires in 2270 days)
  serial:    3d:d6:69:c0:f8:5e:7c:a8:4f:36:4b:d2:2f:5b:87:c6
  flags:     CA CRLSign self-signed
  pathlen:   1
  subjkeyId: 6f:9b:a6:98:bb:59:aa:cd:9f:12:25:70:b1:ac:e2:f5:cb:1f:fa:6d
  pubkey:    RSA 2048 bits
  keyid:     00:98:9c:0b:20:75:a9:7a:28:0f:c0:29:a0:4f:bd:22:bc:05:dc:dc
  subjkey:   6f:9b:a6:98:bb:59:aa:cd:9f:12:25:70:b1:ac:e2:f5:cb:1f:fa:6d

  subject:  "C=CA, ST=QC, O=AAA-BBB, OU=ICP, CN=WEXI800-CA"
  issuer:   "C=CA, ST=QC, O=AAA-BBB, OU=ICP, CN=WEICPACR-CA"
  validity:  not before Sep 07 19:06:46 2023, ok
             not after  Sep 07 19:11:20 2030, ok (expires in 2270 days)
  serial:    53:00:00:00:07:17:8f:d7:70:f9:73:2f:8e:00:02:00:00:00:07
  flags:     CA CRLSign
  pathlen:   0
  authkeyId: 6f:9b:a6:98:bb:59:aa:cd:9f:12:25:70:b1:ac:e2:f5:cb:1f:fa:6d
  subjkeyId: d0:50:9b:f1:5a:09:d3:12:fc:5d:a2:75:f1:75:83:da:4e:ad:fd:bd
  pubkey:    RSA 2048 bits
  keyid:     dd:8c:95:47:5e:1e:f5:fe:d7:4e:8b:64:02:38:1d:90:43:9b:b8:84
  subjkey:   d0:50:9b:f1:5a:09:d3:12:fc:5d:a2:75:f1:75:83:da:4e:ad:fd:bd

----------
Here is the list of certificates on the server’s side:
**ai-lab12_1723**#sh crypto pki certificates
Certificate
  Status: Available
  Certificate Serial Number (hex): 4A00004177706230AA73292903000200004177
  Certificate Usage: General Purpose
  Issuer:
    cn=WEXI800-CA
    ou=ICP
    o=AAA-BBB
    st=QC
    c=CA
  Subject:
    Name: ai-lab12.telecom.aaa.ca
    IP Address: 10.111.222.17 **==> display issue, see the attach image there are 2 addresses in the SAN**
    ou=DPT
    o=AAA-BBB
    l=Lav
    st=QC
    c=CA
  Validity Date:
    start date: 08:22:08 EDT Jun 7 2024
    end   date: 08:32:08 EDT Jun 7 2027

CA Certificate
  Status: Available
  Certificate Serial Number (hex): 53000000141FAB097A38B78EE6000300000014
  Certificate Usage: Signature
  Issuer:
    cn=WEICPACR-CA
    ou=ICP
    o=AAA-BBB
    st=QC
    c=CA
  Subject:
    cn=WEXI800-CA
    ou=ICP
    o=AAA-BBB
    st=QC
    c=CA
  Validity Date:
    start date: 17:03:41 EDT Jun 6 2024
    end   date: 17:13:41 EDT Jun 6 2034

CA Certificate
  Status: Available
  Certificate Serial Number (hex): 64F76BD00243338B466C955D20937F89
  Certificate Usage: Signature
  Issuer:
    cn=WEICPACR-CA
    ou=ICP
    o=AAA-BBB
    st=QC
    c=CA
  Subject:
    cn=WEICPACR-CA
    ou=ICP
    o=AAA-BBB
    st=QC
    c=CA
  Validity Date:
    start date: 09:09:31 EDT Mar 28 2019
    end   date: 17:13:41 EDT Jun 6 2044

<img width="434" alt="SAN" src="https://github.com/strongswan/strongswan/assets/173811973/37bbac56-82fa-4a36-a862-be9a49da6893">

 ----------
conn poc-abc
      keyexchange=ikev1
      leftsourceip=10.164.191.9
      leftsubnet=10.164.191.0/24
      leftcert=C1R6G.cer
      leftauth=pubkey
      right=10.111.222.1
      rightsubnet=10.32.88.0/24
      rightid=%any
      rightauth=pubkey
      ike=3des-sha-modp1024,3des-sha-modp2048,aes256-sha-modp1024,aes256-sha-modp2048
      esp=3des-sha-modp1024,3des-sha-modp2048,aes256-sha-modp1024,aes256-sha-modp2048
      auto=add

root@ubuntupoc:~# ipsec up poc-abc
initiating Main Mode IKE_SA poc-abc[1] to 10.111.222.1
generating ID_PROT request 0 [ SA V V V V V ]
sending packet: from 10.0.2.15[500] to 10.111.222.1[500] (340 bytes)
received packet: from 10.111.222.1[500] to 10.0.2.15[500] (100 bytes)
parsed ID_PROT response 0 [ SA V ]
received NAT-T (RFC 3947) vendor ID
selected proposal: IKE:3DES_CBC/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_1024
generating ID_PROT request 0 [ KE No NAT-D NAT-D ]
sending packet: from 10.0.2.15[500] to 10.111.222.1[500] (244 bytes)
received packet: from 10.111.222.1[500] to 10.0.2.15[500] (492 bytes)
parsed ID_PROT response 0 [ KE No CERTREQ CERTREQ V V V V NAT-D NAT-D ]
received Cisco Unity vendor ID
received DPD vendor ID
received unknown vendor ID: f9:77:51:e9:fc:7e:2d:14:a7:f1:5c:92:88:85:2e:d9
received XAuth vendor ID
received cert request for 'C=CA, ST=QC, O=AAA-BBB, OU=ICP, CN=WEXI800-CA'
received cert request for 'C=CA, ST=QC, O=AAA-BBB, OU=ICP, CN=WEICPACR-CA'
local host is behind NAT, sending keep alives
sending cert request for "C=CA, ST=QC, O=AAA-BBB, OU=ICP, CN=WEXI800-CA"
authentication of 'L=WTFRT101, O=CPS, CN=C1R6G' (myself) successful
sending end entity cert "L=WTFRT101, O=CPS, CN=C1R6G"
generating ID_PROT request 0 [ ID CERT SIG CERTREQ ]
sending packet: from 10.0.2.15[4500] to 10.111.222.1[4500] (1548 bytes)
received packet: from 10.111.222.1[4500] to 10.0.2.15[4500] (1780 bytes)
parsed ID_PROT response 0 [ ID CERT SIG ]
received end entity cert "C=CA, ST=QC, L=Lav, O=AAA-BBB, OU=DPT, CN=ai-lab12.telecom.aaa.ca"
**no trusted RSA public key found for 'ai-lab12_1723.telecom.aaa.ca'**
deleting IKE_SA poc-abc[1] between 10.0.2.15[L=WTFRT101, O=CPS, CN=C1R6G]...10.111.222.1[ai-lab12_1723.telecom.aaa.ca]
sending DELETE for IKE_SA poc-abc[1]
generating INFORMATIONAL_V1 request 1639005121 [ HASH D ]
sending packet: from 10.0.2.15[4500] to 10.111.222.1[4500] (84 bytes)
establishing connection 'poc-abc' failed

**Comment by tobiasbrunner** [comment, marked as answer]
Your server claims that its identity is `ai-lab12_1723.telecom.aaa.ca`, but that identity is not contained in the server certificate as subjectAltName extension. Either change the server so it uses a different identity, or add that identity to the certificate.

**Reply by frmro** [reply]
I added its identity to the SAN and it works. Thank you!

**Reply by itbj** [reply]
Cool. Thanks.

**Comment by cchangwen** [comment]
pki `--san`
